### PR TITLE
Channel v5 table

### LINKED ISSUE #409: AIP #61 v5: Sentry POST route for inserting/modifying campaign
From AIP61 Main issue:
> validator: sentry: will have an authenticated POST to modify/submit campaigns
> - check if `campaign.budget >= campaignSpent`; update a key in Redis for the remaining spendable amount
> - get the last approved NewState from the DB and set `totalSpent = lastApprovedNewState.spenders[user]`
> - set `totalDeposited` to total from the Spendable DB entry for this `(channelId, spender)`
> - make sure `sum(campaigns.map(getRemaining)) <= totalDeposited - totalSpent` when updating campaigns; by doing that we ensure that advertisers can't spend more than they have
> - please note that we can't assume `totalSpent` equals the sum of all campaign spendings, because a spending can occur outside of campaign (cases like v4 CLOSE or PAY); so we must always compare remaining funds based on `totalDeposited - lastApprovedNewState.spenders[user]`
> - to close a campaign, just set it's budget to what it's spent so far (so that `remaining == 0`)

**### To-Dos:**
**Logic:**
- [x] Retrieve `campaignSpent` from Redis
- [x] Check if we have reached the budget
- [x] Calculate the remaining spendable amount for the campaign and update it on Redis
- [x] Retrieve the latest new state
- [x] Retrieve the latest Spendable entry from the DB
- [x] Calculate the total remaining  amount for the channel
- [x] ~??? Does the totalRemaining need to be stored in Redis?~ no
- [x] Compare the remaining amount for all campaigns on this channel with totalDeposited - totalSpent and update the campaign on Postgres

**API:**
- [x] Create an authenticated POST route for modifying campaigns `/campaign/{id}/`
    - [x] Create a middleware for checking if a campaign with this ID exists in Postgres
    - [x] Track changes in budget, possibly with a struct that includes all of the editable fields.
- [x] Create an authenticated POST route for creating campaigns `/campaign/
    - [x] Deserialize CreateCampaign struct from the request
    - [x] Check if there is enough remaining deposit (through spendable) for the campaign budget
    - [x] Return a campaign with a randomly generated ID and insert it to Postgres

**Testing:**
- [x] Refactor business logic to make it easily testable
- [x] Additional test cases for creating & modifying campaign
- [x] Test Postgres update functions 
- [x] Test `sentry::routes::campaign::update_campaign::get_delta_budget`

**Additional Tasks:**
- [x] Implement pagination for `get_campaigns_by_channel(...)` 
- [x] **Very important** - when modifying a Campaign make sure to Authenticate the caller AND validate that he can perform the modification (`AuthRequired` & check if Auth.uid = campaign.creator - could be middleware as well)
- [x] Update Spendable entry as stated in AIP 61 `we'll also update the document from the adapter before a user tries to mutate (add/update) campaigns`
  - [x] Campaign creation was updated in #432
  - [x] Campaign modification #434

### PR DESCRIPTION
**Pull request by elpiel**

### CODE REVIEW

File: sentry/src/db/campaign.rs [hunk 1]
```diff
@@ -39,15 +38,20 @@ pub async fn insert_campaign(pool: &DbPool, campaign: &Campaign) -> Result<bool,
 }
 
 /// ```text
-/// SELECT id, channel, creator, budget, validators, title, pricing_bounds, event_submission, ad_units, targeting_rules, created, active_from, active_to FROM campaigns
-/// WHERE id = $1
+/// SELECT campaign.id, creator, budget, validators, title, pricing_bounds, event_submission, ad_units, targeting_rules, campaign.created, active_from, active_to,
```
Comment (simzzz) [inline sentry/src/db/campaign.rs]: should be `campaigns.id` and `campaigns.created`

File: sentry/src/db/campaign.rs [hunk 2]
```diff
@@ -56,15 +60,18 @@ pub async fn fetch_campaign(
 
 // TODO: We might need to use LIMIT to implement pagination
 /// ```text
-/// SELECT id, channel, creator, budget, validators, title, pricing_bounds, event_submission, ad_units, targeting_rules, created, active_from, active_to
-/// FROM campaigns WHERE channel_id = $1
+/// SELECT campaigns.id, creator, budget, validators, title, pricing_bounds, event_submission, ad_units, targeting_rules, created, active_from, active_to,
```
Comment (simzzz) [inline sentry/src/db/campaign.rs]: should be `campaigns.created`

File: sentry/src/db/campaign.rs [hunk 3]
```diff
@@ -74,14 +81,12 @@ pub async fn get_campaigns_by_channel(
 }
 
 /// ```text
-/// UPDATE campaigns SET budget = $1, validators = $2, title = $3, pricing_bounds = $4, event_submission = $5, ad_units = $6, targeting_rules = $7
-/// WHERE id = $8
-/// RETURNING id, channel, creator, budget, validators, title, pricing_bounds, event_submission, ad_units, targeting_rules, created, active_from, active_to
+/// UPDATE campaigns SET budget = $1, validators = $2, title = $3, pricing_bounds = $4, event_submission = $5, ad_units = $6, targeting_rules = $7 FROM campaigns INNER JOIN channels ON campaigns.channel_id=channels.id WHERE campaigns.id = $8  RETURNING campaigns.id, creator, budget, validators, title, pricing_bounds, event_submission, ad_units, targeting_rules, campaign.created, active_from, active_to, channels.leader, channels.follower, channels.guardian, channels.token, channels.nonce
```
Comment (simzzz) [inline sentry/src/db/campaign.rs]: Should be `FROM channels` instead. The comment also uses `INNER JOIN ... ON` while the statement below does not.

File: sentry/src/db/channel.rs [hunk 4]
```diff
@@ -12,72 +11,63 @@ pub async fn get_channel_by_id(
 ) -> Result<Option<Channel>, PoolError> {
     let client = pool.get().await?;
 
-    let select = client.prepare("SELECT id, creator, deposit_asset, deposit_amount, valid_until, targeting_rules, spec, exhausted FROM channels WHERE id = $1 LIMIT 1").await?;
+    let select = client
+        .prepare("SELECT leader, follower, guardian, token FROM channels WHERE id = $1 LIMIT 1")
```
Comment (simzzz) [inline sentry/src/db/channel.rs]: Shouldn't `nonce` also be selected?

**Review by simzzz — CHANGES_REQUESTED** [review]

File: sentry/src/db/channel.rs (hunk 4 above)
Comment (elpiel) [reply sentry/src/db/channel.rs]: Yes, it should. Good catch!
:heavy_check_mark: Fixed

File: sentry/src/db/campaign.rs (hunk 2 above)
Comment (elpiel) [reply sentry/src/db/campaign.rs]: fixed

File: sentry/src/db/campaign.rs (hunk 1 above)
Comment (elpiel) [reply sentry/src/db/campaign.rs]: fixed

File: sentry/src/db/campaign.rs (hunk 3 above)
Comment (elpiel) [reply sentry/src/db/campaign.rs]: I've updated the full query!

**Review by simzzz — APPROVED** [review]

**Comment by elpiel** [comment]
:raised_hands: :raised_hands: :raised_hands:

**Pull request merged**